File: exercises-cpp/james_clarke/ex02_oo_basics/src/Shapes.cpp
#include "Shapes.h"

int Shape::Sides() const {
	return _sides;
}
double Shape::Perimeter() const {
	return _perimeter;
}
double Shape::Area() const {
	return _area;
}
string Shape::Str() const {
	return this->Type() + ' ' + to_string(_sides) + " sides, " + to_string(_perimeter) + " perimeter, " + to_string(_area) + " area";
}
Shape::Shape() {
	_sides = 0;
	_perimeter = 0;
	_area = 0;
}
void Shape::setSides(int sides) {
	_sides = sides;
}
void Shape::setPerimeter(double perimeter) {
	_perimeter = perimeter;
}
void Shape::setArea(double area) {
	_area = area;
}

string Rectangle::RectangleStr = "Rectangle";
Rectangle::Rectangle(double sideLength, double otherSideLength) {
	setSides(4);
	setPerimeter(2 * sideLength + 2 * otherSideLength);
	setArea(sideLength * otherSideLength);
}
string Rectangle::Type() const {
	return RectangleStr;
}

string Square::SquareStr = "Square";
Square::Square(double sideLength) : Rectangle(sideLength, sideLength) {}
string Square::Type() const {
	return SquareStr;
}

string Circle::CircleStr = "Circle";
Circle::Circle(double radius) {
	setSides(1);
	setPerimeter(2 * M_PI * radius);
	setArea(M_PI * radius * radius);
}
string Circle::Type() const {
	return CircleStr;
}

string Triangle::TriangleStr = "Triangle";
Triangle::Triangle(double height, double base) {
	setSides(3);
	setPerimeter(base + 2 * sqrt(height * height + (base * base / 4.0)));
	setArea(0.5 * base * height);
}
string Triangle::Type() const {
	return TriangleStr;
}

File: exercises-python/james_clarke/ex02_oo_basics/Shapes.py
import math


class Shape:
    _type_name = "Shape"

    def __init__(self, number_of_sides: int, total_area: float, perimeter_length: float):
        self.sides = number_of_sides
        self.area = total_area
        self.perimeter = perimeter_length

    def to_string(self) -> str:
        return self._type_name + " " + str(self.sides) + " sides, " + str(self.perimeter) + " perimeter, " \
            + str(self.area) + " area"


class Rectangle(Shape):
    _type_name = "Rectangle"

    def __init__(self, side_length: float, other_side_length: float):
        my_area = side_length * other_side_length
        my_perimeter = 2 * side_length + 2 * other_side_length
        Shape.__init__(self, number_of_sides=4, total_area=my_area, perimeter_length=my_perimeter)


class Square(Rectangle):
    _type_name = "Square"

    def __init__(self, every_side_length: float):
        Rectangle.__init__(self, side_length=every_side_length, other_side_length=every_side_length)


class Circle(Shape):
    _type_name = "Circle"

    def __init__(self, radius: float):
        my_area = math.pi * radius**2
        my_perimeter = 2 * math.pi * radius
        Shape.__init__(self, number_of_sides=1, total_area=my_area, perimeter_length=my_perimeter)


class Triangle(Shape):
    _type_name = "Triangle"

    def __init__(self, height: float, base: float):
        my_area = 0.5 * base * height
        my_perimeter = base + 2*math.sqrt(height**2 + base**2 / 4)
        Shape.__init__(self, number_of_sides=3, total_area=my_area, perimeter_length=my_perimeter)